File: extensions/pi-cursor/rpc-session.js
"use strict";

// Persistent `pi --mode rpc` session: JSON-lines commands on stdin, responses
// and AgentSessionEvents on stdout. A `prompt` response is only the preflight
// ack; completion is the `agent_end` event (willRetry=false), and provider
// errors surface as the final assistant message's stopReason/errorMessage.

const path = require("path");
const { execFile, spawn } = require("child_process");

const COMMAND_TIMEOUT_MS = 30000;
const PROMPT_TIMEOUT_MS = 180000;

// Windows 没有进程组信号语义：child.kill() 走 TerminateProcess，只终结
// cmd.exe 本身，被它包起来的 pi 会残留成僵尸进程（pi 常以
// %APPDATA%\npm\pi.cmd 安装，cmd 包装是常规路径而非边缘情况）。
// 每次空闲回收 / Key 轮换 respawn 都会漏一个常驻 Node 进程，各自持有一份
// API Key 环境变量。用系统 taskkill /T /F 终结整棵进程树。
// 固定绝对路径 + argv 数组 + 无 shell，与全仓库的 execFile 约定一致。
function defaultKillTree(pid) {
  const systemRoot = process.env.SystemRoot || process.env.windir || "C:\\Windows";
  try {
    execFile(
      path.join(systemRoot, "System32", "taskkill.exe"),
      ["/pid", String(pid), "/T", "/F"],
      { windowsHide: true, timeout: 10000 },
      () => {}
    );
  } catch {
    // 进程树终结是尽力而为，失败时仍有下面的 child.kill() 兜底
  }
}

function extractMessageText(message) {
  if (!message) return "";
  const content = message.content;
  if (typeof content === "string") return content;
  if (!Array.isArray(content)) return "";
  return content
    .filter((block) => block && block.type === "text" && typeof block.text === "string")
    .map((block) => block.text)
    .join("");
}

function lastAssistantMessage(messages) {
  if (!Array.isArray(messages)) return null;
  for (let index = messages.length - 1; index >= 0; index -= 1) {
    const message = messages[index];
    if (message && message.role === "assistant") return message;
  }
  return null;
}

function sessionDeadError(exitInfo, stderrTail) {
  const detail = stderrTail ? `：${stderrTail.slice(-400)}` : "";
  const error = new Error(`Pi RPC 会话已退出（${exitInfo || "未知原因"}）${detail}`);
  error.sessionDead = true;
  // Death before the session ever answered means `--mode rpc` (or pi itself)
  // is unusable in this environment — callers fall back to one-shot mode.
  return error;
}

class PiRpcSession {
  constructor({
    executable,
    args = [],
    env,
    cwd,
    spawnFn = spawn,
    platform = process.platform,
    killTree = defaultKillTree,
  } = {}) {
    this._pending = new Map();
    this._nextId = 1;
    this._buffer = "";
    this._stderrTail = "";
    this._alive = true;
    this._everResponded = false;
    this._exitInfo = null;
    this._turn = null;
    // 每轮请求的单调序号：事件回调只认「当前有效轮次」，超时/中止会让序号
    // 前进从而作废旧轮次。注意 agent_end 事件本身**不带 id**，无法从消息
    // 内容分辨它属于哪一轮，所以序号只能作废「turn 对象」，不能识别陈旧
    // 消息 —— 真正杜绝陈旧事件完成下一轮请求，靠的是超时即销毁会话。
    this._turnSeq = 0;
    this._timedOut = false;
    this._platform = platform;
    this._killTree = killTree;

    this._child = spawnFn(executable, args, {
      cwd,
      env,
      stdio: ["pipe", "pipe", "pipe"],
      windowsHide: true,
    });
    this._child.on("error", (error) => this._onExit(`启动失败: ${error.message}`));
    this._child.on("exit", (code, signal) =>
      this._onExit(`exit code=${code === null ? String(signal) : code}`)
    );
    if (this._child.stdout) {
      this._child.stdout.setEncoding("utf8");
      this._child.stdout.on("data", (chunk) => this._onStdout(chunk));
    }
    if (this._child.stderr) {
      this._child.stderr.setEncoding("utf8");
      this._child.stderr.on("data", (chunk) => {
        this._stderrTail = (this._stderrTail + chunk).slice(-4000);
      });
    }
    // Node 的流错误是**异步 'error' 事件**，不是同步抛出——send() 里的
    // try/catch 接不住。子进程已退出而 write() 仍在途中时 stdin 会 emit
    // 'error'（EPIPE / ERR_STREAM_DESTROYED），没有监听器就是未捕获异常，
    // 足以打崩整个扩展宿主。错误已由 child.on("exit") 统一转成
    // sessionDeadError，此处静默即可。
    if (this._child.stdin && typeof this._child.stdin.on === "function") {
      this._child.stdin.on("error", () => {});
    }
  }

  isAlive() {
    return this._alive;
  }

  isBusy() {
    return this._turn !== null;
  }

  timedOut() {
    return this._timedOut;
  }

  everResponded() {
    return this._everResponded;
  }

  dispose() {
    if (!this._alive) return;
    const pid = this._child ? this._child.pid : undefined;
    if (this._platform === "win32" && Number.isInteger(pid) && pid > 0) {
      this._killTree(pid);
    }
    try {
      this._child.kill();
    } catch {
      // already gone
    }
    this._onExit("disposed");
  }

  _onExit(info) {
    if (!this._alive) return;
    this._alive = false;
    this._exitInfo = info;
    const error = sessionDeadError(info, this._stderrTail);
    if (!this._everResponded) error.rpcUnavailable = true;
    for (const pending of this._pending.values()) {
      clearTimeout(pending.timer);
      pending.reject(error);
    }
    this._pending.clear();
    if (this._turn) {
      const turn = this._turn;
      this._turn = null;
      clearTimeout(turn.timer);
      turn.reject(error);
    }
  }

  _onStdout(chunk) {
    // 进程已死后到达的缓冲数据一律丢弃：不得再驱动任何 turn 完成。
    if (!this._alive) return;
    this._buffer += chunk;
    let newline;
    while ((newline = this._buffer.indexOf("\n")) >= 0) {
      const line = this._buffer.slice(0, newline).trim();
      this._buffer = this._buffer.slice(newline + 1);
      if (!line.startsWith("{")) continue;
      let message;
      try {
        message = JSON.parse(line);
      } catch {
        continue;
      }
      this._handleMessage(message);
    }
  }

  _handleMessage(message) {
    if (!message || typeof message !== "object") return;
    if (message.type === "response") {
      this._everResponded = true;
      const pending = this._pending.get(String(message.id || ""));
      if (pending) {
        this._pending.delete(String(message.id || ""));
        clearTimeout(pending.timer);
        pending.resolve(message);
      }
      return;
    }
    const turn = this._turn;
    // 只接受「当前有效轮次」的事件：turn 被超时作废后 _turnSeq 已前进，
    // 任何残留事件都无法再完成一轮请求。
    if (!turn || turn.seq !== this._turnSeq) return;
    if (message.type === "message_end" && message.message && message.message.role === "assistant") {
      turn.lastAssistant = message.message;
      return;
    }
    if (message.type === "agent_end" && !message.willRetry) {
      const fromEnd = lastAssistantMessage(message.messages);
      if (fromEnd) turn.lastAssistant = fromEnd;
      this._turn = null;
      clearTimeout(turn.timer);
      turn.resolve();
    }
  }

  send(command, timeoutMs = COMMAND_TIMEOUT_MS) {
    if (!this._alive) {
      const error = sessionDeadError(this._exitInfo, this._stderrTail);
      if (!this._everResponded) error.rpcUnavailable = true;
      return Promise.reject(error);
    }
    const id = String(this._nextId++);
    const payload = { ...command, id };
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this._pending.delete(id);
        reject(new Error(`Pi RPC 命令超时：${String(command.type || "")}`));
      }, timeoutMs);
      this._pending.set(id, { resolve, reject, timer });
      try {
        this._child.stdin.write(`${JSON.stringify(payload)}\n`);
      } catch (error) {
        this._pending.delete(id);
        clearTimeout(timer);
        reject(error);
      }
    });
  }

  async setModel(provider, modelId) {
    const response = await this.send({ type: "set_model", provider, modelId });
    if (!response || response.success === false) {
      throw new Error(
        `切换模型失败：${(response && response.error) || "未知错误"}`
      );
    }
    return response.data || null;
  }

  async prompt(text, { timeoutMs = PROMPT_TIMEOUT_MS } = {}) {
    if (this._turn) throw new Error("上一个 Pi 请求仍在进行");
    const started = Date.now();
    const seq = (this._turnSeq += 1);
    const timeoutError = () =>
      new Error(`Pi 响应超时（${Math.round(timeoutMs / 1000)}s）`);
    const turnDone = new Promise((resolve, reject) => {
      this._turn = {
        seq,
        lastAssistant: null,
        resolve,
        reject,
        timer: setTimeout(() => {
          const turn = this._turn;
          if (!turn || turn.seq !== seq) return;
          // 先作废本轮（_turnSeq 前进使残留事件失配），再礼貌地发 abort，
          // 最后**销毁会话**。abort 是 fire-and-forget：子进程可能仍在处理，
          // 稍后才吐出本轮的 agent_end；该事件不带 id，若会话留活就会被下
          // 一轮请求当成自己的完成信号，在毫秒级内"成功返回"上一轮的答案，
          // 并被 failover 记为成功、重置失败计数（审查报告 P2-5）。
          // sticky --session-id 让下一次提问重建同一会话，上下文照样恢复，
          // 代价只是一次进程启动。
          // 超时用 resolve 而不是 reject：时钟可能在 await send() 期间响，
          // 那时还没有人 await turnDone，reject 会变成 unhandledRejection
          // 打崩扩展宿主（CI Linux runner 上 30ms 窗口即可复现）。
          this._turn = null;
          this._turnSeq += 1;
          this._timedOut = true;
          this.send({ type: "abort" }).catch(() => {});
          turn.resolve({ timedOut: true });
          this.dispose();
        }, timeoutMs),
      };
    });
    const turn = this._turn;

    const finish = (patch) => ({
      ok: false,
      returncode: -1,
      stdout: "",
      stderr: "",
      latency_ms: Date.now() - started,
      error: "",
      ...patch,
    });

    let ack;
    try {
      ack = await this.send({ type: "prompt", message: String(text) });
    } catch (error) {
      if (this._turn === turn) {
        this._turn = null;
        clearTimeout(turn.timer);
      }
      if (this._timedOut) return finish({ error: timeoutError().message });
      if (error.sessionDead) throw error;
      return finish({ error: error.message });
    }
    if (!ack || ack.success === false) {
      if (this._turn === turn) {
        this._turn = null;
        clearTimeout(turn.timer);
      }
      return finish({ error: (ack && ack.error) || "prompt 预检失败" });
    }

    try {
      const outcome = await turnDone;
      if (outcome && outcome.timedOut) {
        return finish({ error: timeoutError().message });
      }
    } catch (error) {
      if (this._timedOut) return finish({ error: timeoutError().message });
      if (error.sessionDead) throw error;
      return finish({ error: error.message });
    }

    const message = turn.lastAssistant;
    const stopReason = message && message.stopReason;
    if (stopReason === "error" || stopReason === "aborted") {
      return finish({
        error:
          (message && message.errorMessage) ||
          `模型返回 ${stopReason === "aborted" ? "已中止" : "错误"}`,
      });
    }

    let answer = "";
    try {
      const response = await this.send({ type: "get_last_assistant_text" });
      if (response && response.success !== false && response.data) {
        answer = String(response.data.text || "");
      }
    } catch {
      // fall back to the streamed message below
    }
    if (!answer) answer = extractMessageText(message);
    if (!answer.trim()) return finish({ error: "模型没有返回文本" });
    return finish({ ok: true, returncode: 0, stdout: answer, error: "" });
  }
}

module.exports = {
  PiRpcSession,
  extractMessageText,
  lastAssistantMessage,
};

File: pi_manager/rpc_session.py
"""Persistent `pi --mode rpc` session for the desktop quick chat.

One long-lived pi process holds the conversation; model changes are applied
hot via set_model (context preserved in-process). Credential changes require
a respawn because env cannot be injected into a running child — the sticky
--session-id makes pi reload the same session file, so context survives
restarts too.
"""
from __future__ import annotations

import json
import logging
import subprocess
import threading
import time
import uuid
from collections.abc import Callable
from typing import Any

from . import core
from . import proc
from . import secrets as secretstore

logger = logging.getLogger(__name__)

_COMMAND_TIMEOUT = 30.0
_PROMPT_TIMEOUT = 180.0
_RUNTIME_RETRY_COOLDOWN = 30.0
_MAX_STDOUT_LINE = 10 * 1024 * 1024
_STDIN_WRITE_TIMEOUT = 10.0
_STDIN_WRITE_GRACE = 0.2
_MAX_PROVIDER_ENV_CACHE = 8


class RpcSessionError(RuntimeError):
    def __init__(self, message: str, *, unavailable: bool = False) -> None:
        super().__init__(message)
        self.unavailable = unavailable


def _extract_text(message: dict[str, Any] | None) -> str:
    if not isinstance(message, dict):
        return ""
    content = message.get("content")
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ""
    return "".join(
        str(block.get("text") or "")
        for block in content
        if isinstance(block, dict) and block.get("type") == "text"
    )


class PiRpcSession:
    def __init__(self, argv: list[str], *, env: dict[str, str], cwd: str | None = None) -> None:
        creationflags = proc.create_no_window_flag()
        self._proc = subprocess.Popen(
            argv,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            cwd=cwd or None,
            env=env,
            text=True,
            encoding="utf-8",
            errors="replace",
            creationflags=creationflags,
        )
        self._stdin_lock = threading.Lock()
        self._state_lock = threading.Lock()
        self._pending: dict[str, dict[str, Any]] = {}
        self._turn: dict[str, Any] | None = None
        self._next_id = 1
        self._alive = True
        self._ever_responded = False
        self._exit_info = ""
        self._stderr_tail = ""
        threading.Thread(target=self._read_stdout, daemon=True).start()
        threading.Thread(target=self._read_stderr, daemon=True).start()

    def is_alive(self) -> bool:
        with self._state_lock:
            return self._alive

    def is_busy(self) -> bool:
        with self._state_lock:
            return self._turn is not None

    def close(self) -> None:
        try:
            if self._proc.poll() is None:
                self._proc.kill()
        except Exception as exc:
            logger.warning("关闭 Pi RPC 进程失败: %s", exc)
        self._on_exit("closed")

    def _read_stdout(self) -> None:
        try:
            for raw in self._proc.stdout or []:
                line = raw.strip()
                if not line.startswith("{"):
                    continue
                if len(line) > _MAX_STDOUT_LINE:
                    logger.warning("Pi RPC stdout 行过长 (%d 字符)，跳过", len(line))
                    continue
                try:
                    message = json.loads(line)
                except ValueError:
                    continue
                self._handle(message)
        except Exception as exc:
            logger.warning("读取 Pi RPC stdout 失败: %s", exc)
        self._on_exit(f"exit code={self._proc.poll()}")

    def _read_stderr(self) -> None:
        try:
            for line in self._proc.stderr or []:
                self._stderr_tail = (self._stderr_tail + line)[-4000:]
        except Exception as exc:
            logger.warning("读取 Pi RPC stderr 失败: %s", exc)

    def _handle(self, message: Any) -> None:
        if not isinstance(message, dict):
            return
        if message.get("type") == "response":
            with self._state_lock:
                self._ever_responded = True
                pending = self._pending.pop(str(message.get("id") or ""), None)
            if pending is not None:
                pending["response"] = message
                pending["event"].set()
            return
        with self._state_lock:
            turn = self._turn
        if not turn:
            return
        if message.get("type") == "message_end":
            inner = message.get("message")
            if isinstance(inner, dict) and inner.get("role") == "assistant":
                with self._state_lock:
                    if self._turn is turn:
                        turn["last_assistant"] = inner
            return
        if message.get("type") == "agent_end" and not message.get("willRetry"):
            for inner in reversed(message.get("messages") or []):
                if isinstance(inner, dict) and inner.get("role") == "assistant":
                    with self._state_lock:
                        if self._turn is turn:
                            turn["last_assistant"] = inner
                    break
            with self._state_lock:
                if self._turn is turn:
                    self._turn = None
            turn["event"].set()

    def _on_exit(self, info: str) -> None:
        with self._state_lock:
            if not self._alive:
                return
            self._alive = False
            self._exit_info = info
            pendings = list(self._pending.values())
            self._pending.clear()
            turn = self._turn
            self._turn = None
        for pending in pendings:
            pending["event"].set()
        if turn is not None:
            turn["dead"] = True
            turn["event"].set()

    def _dead_error(self) -> RpcSessionError:
        detail = f"：{self._stderr_tail[-400:]}" if self._stderr_tail else ""
        return RpcSessionError(
            f"Pi RPC 会话已退出（{self._exit_info or '未知原因'}）{detail}",
            unavailable=not self._ever_responded,
        )

    def _write_payload(self, payload: dict[str, Any]) -> None:
        """Write one JSON line to the child stdin with a bounded flush.

        A wedged child that stops reading stdin would otherwise block
        write()/flush() forever and bypass the command timeout. The write
        runs on a worker thread; on timeout the child is killed (a child
        that cannot read stdin cannot serve further RPC anyway) and the
        caller receives a normal RpcSessionError instead of hanging.
        """
        line = json.dumps(payload, ensure_ascii=False) + "\n"
        with self._stdin_lock:
            if self._proc.stdin is None:
                raise RpcSessionError("Pi RPC stdin 不可用")
            outcome: dict[str, Any] = {}

            def do_write() -> None:
                try:
                    assert self._proc.stdin is not None
                    self._proc.stdin.write(line)
                    self._proc.stdin.flush()
                    outcome["ok"] = True
                except Exception as exc:  # BrokenPipeError / OSError ...
                    outcome["exc"] = exc

            worker = threading.Thread(target=do_write, daemon=True)
            worker.start()
            worker.join(_STDIN_WRITE_TIMEOUT)
            if worker.is_alive():
                # Grace tick: the worker may have just finished its flush but
                # not yet returned; only declare a timeout if still stuck.
                worker.join(_STDIN_WRITE_GRACE)
            if worker.is_alive():
                self.close()
                raise RpcSessionError(
                    f"写入 Pi RPC stdin 超时（{_STDIN_WRITE_TIMEOUT:.0f}s），子进程已终止"
                )
            if "exc" in outcome:
                raise RpcSessionError(
                    f"发送 RPC 命令失败：{outcome['exc']}"
                ) from outcome["exc"]

    def send(self, command: dict[str, Any], timeout: float = _COMMAND_TIMEOUT) -> dict[str, Any]:
        with self._state_lock:
            if not self._alive:
                raise self._dead_error()
            command_id = str(self._next_id)
            self._next_id += 1
            pending: dict[str, Any] = {"event": threading.Event(), "response": None}
            self._pending[command_id] = pending
        payload = dict(command)
        payload["id"] = command_id
        try:
            self._write_payload(payload)
        except Exception as exc:
            with self._state_lock:
                self._pending.pop(command_id, None)
            raise RpcSessionError(f"发送 RPC 命令失败：{exc}") from exc
        if not pending["event"].wait(timeout):
            with self._state_lock:
                self._pending.pop(command_id, None)
            raise RpcSessionError(f"Pi RPC 命令超时：{command.get('type')}")
        if pending["response"] is None:
            raise self._dead_error()
        return pending["response"]

    def set_model(self, provider: str, model_id: str) -> Any:
        response = self.send({"type": "set_model", "provider": provider, "modelId": model_id})
        if response.get("success") is False:
            raise RpcSessionError(f"切换模型失败：{response.get('error') or '未知错误'}")
        return response.get("data")

    def prompt(
        self,
        text: str,
        timeout: float = _PROMPT_TIMEOUT,
        *,
        is_cancelled: Callable[[], bool] | None = None,
    ) -> dict[str, Any]:
        with self._state_lock:
            if self._turn is not None:
                raise RpcSessionError("上一个 Pi 请求仍在进行")
            turn: dict[str, Any] = {
                "event": threading.Event(),
                "last_assistant": None,
                "dead": False,
            }
            self._turn = turn
        started = time.perf_counter()

        def finish(**patch: Any) -> dict[str, Any]:
            result: dict[str, Any] = {
                "ok": False,
                "returncode": -1,
                "stdout": "",
                "stderr": "",
                "latency_ms": round((time.perf_counter() - started) * 1000, 1),
                "error": "",
            }
            result.update(patch)
            return result

        def clear_turn() -> None:
            with self._state_lock:
                if self._turn is turn:
                    self._turn = None

        try:
            ack = self.send({"type": "prompt", "message": str(text)})
        except RpcSessionError:
            clear_turn()
            raise
        if ack.get("success") is False:
            clear_turn()
            return finish(error=str(ack.get("error") or "prompt 预检失败"))

        deadline = time.monotonic() + max(0.05, float(timeout))
        cancelled = False
        while not turn["event"].is_set():
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                break
            if is_cancelled and is_cancelled():
                cancelled = True
                break
            turn["event"].wait(min(0.15, remaining))
        if cancelled or not turn["event"].is_set():
            clear_turn()
            try:
                self.send({"type": "abort"})
            except RpcSessionError:
                pass
            if cancelled:
                return finish(error="已停止生成", cancelled=True)
            return finish(error=f"Pi 响应超时（{int(timeout)}s）")
        if turn.get("dead"):
            raise self._dead_error()

        message = turn.get("last_assistant") or {}
        stop_reason = message.get("stopReason")
        if stop_reason in ("error", "aborted"):
            return finish(
                error=str(
                    message.get("errorMessage")
                    or ("模型返回已中止" if stop_reason == "aborted" else "模型返回错误")
                )
            )

        answer = ""
        try:
            response = self.send({"type": "get_last_assistant_text"})
            if response.get("success") is not False:
                answer = str((response.get("data") or {}).get("text") or "")
        except RpcSessionError:
            pass
        if not answer:
            answer = _extract_text(message)
        if not answer.strip():
            return finish(error="模型没有返回文本")
        return finish(ok=True, returncode=0, stdout=answer, error="")


# ---- desktop chat session manager -----------------------------------------

_manager_lock = threading.Lock()
_entry: dict[str, Any] | None = None
_runtime_disabled = False
_runtime_disabled_since = 0.0
_idle_timer: threading.Timer | None = None
# 专用于 _runtime_disabled / _runtime_disabled_since 的跨线程同步：
# 这两个变量由 Worker 线程写、主线程读，必须成对一致读取，避免快照错位。
_runtime_lock = threading.Lock()


def _idle_ttl_seconds() -> float:
    """Idle minutes before the persistent pi process is reclaimed (0 = never)."""
    try:
        minutes = float(core.load_manager_config().get("chat_session_idle_min") or 10)
    except Exception:
        minutes = 10.0
    return max(0.0, minutes) * 60.0


def _schedule_idle_reaper() -> None:
    """(Re)arm the idle reaper. Caller must hold _manager_lock."""
    global _idle_timer
    if _idle_timer is not None:
        _idle_timer.cancel()
        _idle_timer = None
    ttl = _idle_ttl_seconds()
    if ttl <= 0 or _entry is None:
        return
    timer = threading.Timer(ttl, _reap_idle_session)
    timer.daemon = True
    _idle_timer = timer
    timer.start()


def _reap_idle_session() -> None:
    """Close the persistent session once it has sat idle for a full TTL.

    The sticky --session-id means a later prompt transparently reloads the
    same conversation, so reclaiming the process costs nothing but latency.
    """
    global _entry, _idle_timer
    with _manager_lock:
        entry = _entry
        _idle_timer = None
        if entry is None:
            return
        session = entry["session"]
        if session.is_busy() or time.monotonic() - entry.get("last_used", 0.0) < _idle_ttl_seconds() - 1:
            _schedule_idle_reaper()
            return
        _entry = None
    session.close()


def rpc_chat_enabled() -> bool:
    global _runtime_disabled
    with _runtime_lock:
        if _runtime_disabled:
            if time.monotonic() - _runtime_disabled_since < _RUNTIME_RETRY_COOLDOWN:
                return False
            _runtime_disabled = False
    try:
        mgr = core.load_manager_config()
    except Exception:
        return False
    return bool(mgr.get("chat_persistent_session", True))


def reset_chat_session() -> None:
    """Drop the persistent conversation (new session id on next prompt)."""
    global _entry, _idle_timer
    with _manager_lock:
        entry = _entry
        _entry = None
        if _idle_timer is not None:
            _idle_timer.cancel()
            _idle_timer = None
    if entry is not None:
        entry["session"].close()


def _ensure(
    provider: str,
    model: str,
    provider_env: dict[str, str],
    workdir: str | None,
    thinking: str,
) -> dict[str, Any]:
    global _entry
    entry = _entry
    workdir_key = str(workdir or "")
    env_snapshot = dict(provider_env)
    # A1: keep one env snapshot per provider so the env comparison is precise.
    # The child process is started with a fixed --provider, so a switch to a
    # different provider must respawn regardless of the cached env (pi's RPC
    # set_model cannot find models of another provider); within the same
    # provider, respawn only when its env genuinely changed (e.g. key
    # rotation), never because only the last provider's env was remembered.
    env_changed = entry is not None and entry["env_by_provider"].get(provider) != env_snapshot
    respawn = (
        entry is None
        or not entry["session"].is_alive()
        or entry["current"][0] != provider
        or env_changed
        or entry["workdir"] != workdir_key
        or entry["thinking"] != thinking
    )
    if respawn:
        session_id = entry["session_id"] if entry else str(uuid.uuid4())
        base = core.pi_base_cmd()
        argv = base + core.escape_cmd_shim_args(
            ["--mode", "rpc", "--provider", provider, "--model", model],
            base,
        )
        if thinking:
            # escape_cmd_shim_args 只校验传入的 provider/model 列表，thinking 在
            # 转义之后单独 append 会绕过白名单，成为 argv 参数注入面（thinking 值
            # 可经 settings.json 的 defaultThinkingLevel 被配置包导入控制）。此处
            # 显式校验后再拼接，非法值直接抛 ValueError 拒绝启动。
            core.validate_launch_tokens(["--thinking", thinking])
            argv += ["--thinking", thinking]
        argv += ["--session-id", session_id, "-n", "PiManager 快速提问"]
        spawn_env = proc.spawn_env(provider_env, sanitize_after_merge=False)
        # A2: spawn the replacement first; only tear the old session down
        # after the new process is up, so a spawn failure keeps the old
        # conversation recoverable instead of losing it.
        session = PiRpcSession(argv, env=spawn_env, cwd=workdir or None)
        if entry is not None:
            entry["session"].close()
        # A1: carry over the per-provider env cache and bound its size so the
        # cache cannot grow without limit across provider switches.
        env_by_provider = dict(entry["env_by_provider"]) if entry else {}
        env_by_provider[provider] = env_snapshot
        if len(env_by_provider) > _MAX_PROVIDER_ENV_CACHE:
            for stale in list(env_by_provider):
                if stale == provider:
                    continue
                del env_by_provider[stale]
                if len(env_by_provider) <= _MAX_PROVIDER_ENV_CACHE:
                    break
        _entry = {
            "session": session,
            "session_id": session_id,
            "env_by_provider": env_by_provider,
            "current": (provider, model),
            "workdir": workdir_key,
            "thinking": thinking,
        }
        return _entry
    if entry["current"] != (provider, model):
        entry["session"].set_model(provider, model)
        entry["current"] = (provider, model)
    return entry


def _failed(provider: str | None, model: str | None, error: str) -> dict[str, Any]:
    return {
        "ok": False,
        "returncode": -1,
        "stdout": "",
        "stderr": "",
        "latency_ms": 0,
        "provider": provider,
        "model": model,
        "error": error,
    }


def rpc_chat_once(
    prompt: str,
    *,
    provider: str | None = None,
    model: str | None = None,
    workdir: str | None = None,
    timeout: float = 180,
    thinking: str | None = "off",
    is_cancelled: Callable[[], bool] | None = None,
) -> dict[str, Any]:
    """chat_once-shaped attempt over the persistent RPC session.

    Same key-rotation semantics as run_pi_print: auth/limit failures mark the
    managed key and retry with the next one (respawn keeps the conversation
    via the sticky session id); other failures are returned for the model
    failover layer to count.
    """
    global _entry, _runtime_disabled, _runtime_disabled_since
    if not provider or not model:
        return _failed(provider, model, "Provider 和 Model 必须成对指定")
    if _runtime_disabled and time.monotonic() - _runtime_disabled_since >= _RUNTIME_RETRY_COOLDOWN:
        with _runtime_lock:
            _runtime_disabled = False
    attempted: set[str] = set()
    last: dict[str, Any] | None = None
    while True:
        try:
            credential = core.provider_runtime_credential(provider)
        except Exception as exc:
            return _failed(provider, model, str(exc))
        key_id = str(credential.get("key_id") or "")
        if key_id and key_id in attempted:
            return last or _failed(provider, model, "API Key 轮换未提供新的可用 Key")
        try:
            with _manager_lock:
                entry = _ensure(
                    provider, model, dict(credential.get("env") or {}), workdir, str(thinking or "off")
                )
            session = entry["session"]
            result = session.prompt(prompt, timeout=timeout, is_cancelled=is_cancelled)
        except RpcSessionError as exc:
            with _manager_lock:
                if _entry is not None and not _entry["session"].is_alive():
                    _entry = None
            if exc.unavailable:
                with _runtime_lock:
                    _runtime_disabled = True
                    _runtime_disabled_since = time.monotonic()
            result = _failed(provider, model, str(exc))
        except FileNotFoundError as exc:
            with _runtime_lock:
                _runtime_disabled = True
                _runtime_disabled_since = time.monotonic()
            return _failed(provider, model, str(exc))
        with _manager_lock:
            if _entry is not None:
                _entry["last_used"] = time.monotonic()
                _schedule_idle_reaper()
        result["provider"], result["model"] = provider, model
        if result.get("cancelled"):
            return result
        if result.get("ok") or not key_id:
            if result.get("ok"):
                with _runtime_lock:
                    _runtime_disabled = False
            return result
        classification = core.classify_provider_key_failure(
            int(result.get("returncode") or -1),
            str(result.get("stdout") or ""),
            str(result.get("error") or result.get("stderr") or ""),
        )
        if not classification.get("status"):
            return result
        try:
            changed = secretstore.mark_provider_key_failed(
                provider, key_id, str(result.get("error") or "")[:200]
            )
        except Exception:
            return result
        if not changed:
            return result
        attempted.add(key_id)
        last = result
